force-app/main/default/classes/DMLOperations.cls

```diff
@@ -180,28 +256,78 @@ public with sharing class DMLOperations {
 	 * @param contactsToUpsert List of Contacts to be upserted and linked to accounts.
 	 */
 	public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
-		// Loop through each Contact
-
-		// Extract the Account Name from Contact's LastName
-
+		// List to hold each contacts last name to be used as the Account Name
+        List<String> accountNames = new List<String>();
+        // Loop through each Contact
+        for (Contact con : contactsToUpsert) {
+            // Extract the Account Name from Contact's LastName
+            String accountName = con.LastName;
+            // Add the Account Name to the accountNames list
+            accountNames.add(accountName);
+        }
 		// Get the account based on the name or create one if it doesn't exist
+        List<Account> accounts = [SELECT Id, Name FROM Account WHERE Name IN :accountNames];
+        // Loop through all the accountNames and check if there is an account with that name if not create one
+        for (Integer i = 0, len = accountNames.size(); i < len; i++) {
+            Boolean found = false;
+            // Compare the current Account Name to all of the Accounts in the List that was queried
+            for (Account acc : accounts) {
+                if (acc.Name.compareTo(accountNames[i]) == 0) {
+                    // If the Account is found, set found to true and break the loop
+                    found = true;
+                    break;
+                }
+            }
+            // If the accountName is not found in the accounts List, then create the Account and add it to the accounts List
+            if (found == false) {
+                Account newAcc = new Account(Name = accountNames[i]);
+                accounts.add(newAcc);
+            }
+        }
+        // Upsert the Accounts
+        upsert accounts;
+        // Loop through each Contact in contactsToUpsert List
+        for (Contact con : contactsToUpsert) {
+            // Loop through each account in the accounts List and compare the Account.Name to the Contact.LastName
+            for (Account acc : accounts) {
+                if (acc.Name.compareTo(con.LastName) == 0) {
+                    // If the Account.Name and the Contact.LastName match then set the AccountId of the Contact to the matching Account Id
+                    con.AccountId = acc.Id;
+                    break;
+                }
+            }
+        }
+        // Upsert the Contacts
+        upsert contactsToUpsert;
```

Comment:
This was a tricky problem. In a real production database I would say this method was too long and does too many things. If you made use of the the upsertAccount method you could do this is a lot less code.  Although I grant that the home solution would not be the best practice in production either.

Reply:
@bdJohnson72 after doing HW5 I just realized that I could save some time by calling previous methods in the Class to help me with the concept of DRY. I'm going to have to give this a shot in future homeworks!